qiskit/transpiler/passes/optimization/commutative_cancellation.py

```diff
@@ -147,22 +149,26 @@ def run(self, dag):
 
                     # Compose gates
                     total_angle = current_angle + total_angle
+                    if current_node.op.definition:
+                        total_phase += current_node.op.definition.global_phase
```

Comment:
Is this always needed (even if we don't need to unroll `current_node`)?

Reply:
If preservation of global phase can be evaluated by Operator(before) == Operator(after) then this is needed since Operator descends through gate definitions until it gets to `to_matrix` implementations. Without this line 6 tests will fail.